pkg/operator/controller/typed.go

```diff
@@ -0,0 +1,207 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"net/http"
+	"reflect"
+
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type TypedBuilder interface {
+	Builder
+
+	// For updates the builder to watch the client.Object passed in
+	For(client.Object) TypedBuilder
+}
+
+type TypedBuilderAdapter struct {
+	builder *controllerruntime.Builder
+}
+
+func NewTypedBuilderAdapter(builder *controllerruntime.Builder) *TypedBuilderAdapter {
+	return &TypedBuilderAdapter{
+		builder: builder,
+	}
+}
+
+func (t *TypedBuilderAdapter) For(obj client.Object) TypedBuilder {
+	t.builder = t.builder.For(obj)
+	return t
+}
+
+func (t *TypedBuilderAdapter) Complete(r reconcile.Reconciler) error {
+	return t.builder.Complete(r)
+}
+
+type TypedReconciler[T client.Object] interface {
+	Reconcile(context.Context, T) (T, reconcile.Result, error)
```

Comment:
I'd lean away from additional interfaces beyond this one. You're essentially building a type system into go using type inference. I'd avoid codifying practices that rely on users to spell methods right and skip compile time checks.

Reply:
This could be solved with a registration system, where someone opts in to specific hooks and handler based on those hooks, that would be more explicit and strongly typed, but I get your overall point.

Rationale behind the overall decision to do this below